### ISSUE TITLE
Ability to use bluetooth keyboard to iterate through ankidroid

### ISSUE THREAD

**Issue by RossRogers** [ankidroid/Anki-Android#3600]
I have a bluetooth gamepad/keyboard device ([here](http://www.gearbest.com/cables-adapter/pp_119065.html)) that I'd like to use to advance cards, go back, flip over, and mark answer as correct or incorrect.

Is there a way to enable a keyboard listener in ankidroid to listen for the bluetooth "keyboard's" key press events?

If there isn't a mechanism (I couldn't find one), would such a contribution be accepted to your trunk?

Do you have any pointers about how you'd go about this in your code base?  I've written small android apps before.

**Comment by agrueneberg** [comment]
Some basic shortcuts (modeled after Anki Desktop, i.e. `Space` to unveil answer, `1` to `4` to answer cards, etc.) are already available for regular keyboards. Can you set up your remote control to map to those keys?

**Comment by RossRogers** [comment]
Excellent.  I will look into that shortly.

**Comment by lizzgilmore** [comment]
@RossRogers - If you're still subscribed to this, may I ask, did you ever get that gamepad to work? I haven't found a way to map the buttons on the gamepad to 1-4 or Space.

**Comment by RossRogers** [comment]
I haven't had the time to try it lately.  I do need to get back to this.

**Comment by lizzgilmore** [comment]
No worries. I played around with it a bit more and eventually managed to
sort of remap the gamepad using the $1.99 paid app External Keyboard Helper
Pro. The software is clunky and you have to make sure you manually switch
input method to the one named after the app, but it worked!

**Comment by jhegedus42** [comment]
What kind of gamepad are you using @lizzgilmore? I am looking for something like Zeemote. I want to control Ankidroid for sound only cards, listening to the questions / answers and giving feedback with a remote controller, some kind of button which I can press without needing to look , for example when driving or running or even when relaxing with closed eyes before going to sleep. Zeemote was great for this, any suggestion what might be used for this purpose?

**Comment by RossRogers** [comment]
This is the one I got. [Compact Bluetooth Gamepad Remote Control Camera Shutter Self - Timer  -  BLACK](http://www.gearbest.com/cables-adapter/pp_119065.html)

**Comment by lizzgilmore** [comment]
Yep, that's the one I got, though I bought mine on Amazon... There are multiple separate listings for the same controller, some with Prime shipping: http://www.amazon.com/BEBONCOOL-Wireless-Bluetooth-Controller-Gamepad/dp/B00UJC2KW0/ref=sr_1_1?ie=UTF8&qid=1454171543&sr=8-1-spons&keywords=bluetooth+gamepad&psc=1 vs http://www.amazon.com/Docooler%C2%AE-Multifunction-Bluetooth-Controller-Compatible/dp/B00SCOAL0K/ref=sr_1_9?ie=UTF8&qid=1454171496&sr=8-9&keywords=bluetooth+gamepad

**Comment by jhegedus42** [comment]
Thank you guys for the info. This is exactly the kind of thing I was looking for.

**Comment by lizzgilmore** [comment]
You may also be interested in this post I wrote on how to map the volume keys on your phone to the Anki shortcut of your choosing. 

https://github.com/ankidroid/Anki-Android/issues/3021#issuecomment-160243369

You may decide that using a bluetooth gamepad is more ergonomic, but I also had success using that method to remap the volume buttons on a set of earbuds, which was pretty comfortable to hold.

**Comment by RossRogers** [comment]
Wow!  Earbud controls! Brilliant.

**Comment by timrae** [comment]
I added a link to this thread in the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ#can-i-use-a-hardware-controller)

**Comment by tbo1** [comment]
I tried the way described above (thanks a lot for the information). Actually I set 1 to 4, space and r for repeating audio (I bought a cheap small gamepad for that). But in any case 1 and space does the same (probably "good" or "hard") instead of "fail". Is there actually a problem with the key commands in the recent ankidroid version?

Thanks in advance
Thomas

**Comment by tbo1** [comment]
I found the solution by myself. Anki is fine. My remote control seems to send two commands at the same time (+ space) by pressing the buttons. So I use the cursor instead which works brilliant.
Up=false
Down good
left r (repeat)
right 4 (easy)
which is by the way easy to remember

**Comment by AustinHasten** [comment]
KeyEvent has constants for gamepad buttons, so it should be trivial to add support for gamepads. The only issue I see is imposing a certain button layout for every user. For example, through a remapping app I have:

BUTTON_X mapped to 1 (Again)
BUTTON_Y mapped to 2 (Hard)
BUTTON_A mapped to 3 (Good)
BUTTON_B mapped to 4 (Easy)

I like this scheme, and if it's acceptable I don't see a reason not to add it. Thoughts?

**Comment by AustinHasten** [comment]
I've submitted an improvement to support gamepads here: https://github.com/ankidroid/Anki-Android/issues/4829

**Comment by eschava** [comment]
> KeyEvent has constants for gamepad buttons, so it should be trivial to add support for gamepads. The only issue I see is imposing a certain button layout for every user. For example, through a remapping app I have:
> 
> BUTTON_X mapped to 1 (Again)
> BUTTON_Y mapped to 2 (Hard)
> BUTTON_A mapped to 3 (Good)
> BUTTON_B mapped to 4 (Easy)

Looks like it is a bit different according to https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/Reviewer.java :
KEYCODE_BUTTON_Y - EASE_1
KEYCODE_BUTTON_X - EASE_2
KEYCODE_BUTTON_B - EASE_3
KEYCODE_BUTTON_A - EASE_4

**Comment by AustinHasten** [comment]
@eschava It's been a while,  but I think that's just due to the fact that Android uses the Xbox button layout for those variable names, whereas I was talking about the Nintendo controller I was using :)
